### ISSUE TITLE
[sklearn#34604] RFC Always store fitted attributes as Numpy

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#311]
**Mirrored from [scikit-learn#34604](https://github.com/scikit-learn/scikit-learn/issues/34604)**

Original title: RFC Always store fitted attributes as Numpy
Original labels: Array API
Created: 2026-07-30T13:41:09Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

This is an issue related to #34152 but is also a standalone issue.

While thinking about #34152 and working on #34324 I thought that it might be a good idea if the fitted attributes of an estimator were always Numpy arrays. The current plan is to store fitted attributes in the namespace/device of `X`.

Reasons to do this:
1. users most often use fitted attributes for inspection/debugging/plotting. Having them in a consistent array namespace makes this easy to do. If the namespace is input dependent then you have to change your debugging/plotting code. Say you run the same script on a machine with a CUDA GPU and on your laptop with no GPU. While it is not difficult to use write code that uses `move_to()` to conditionally bring the attributes to Numpy for plotting, it is tedious.
2. Today attributes are Numpy, by keeping them Numpy we remove one source of bugs/differences. Enabling array API by default would not change this
3. Part of #34324 that is not great/tricky is dealing with estimators that sometimes support array API. For example for some hyper-parameter values array API is supported and for some not. The tricky question: What should the type of the fitted attributes be?
4. You can `fit` with one array namespace/device and `predict` with a different one without any conversion
5. An estimator fitted with pytorch/CUDA can be unpickled on a machine without CUDA and even without pytorch

Something to check is how much overhead/performance loss we get in `predict` and other inference methods if we have to convert the fitted attributes relevant for predicting from Numpy to the array namespace of the input each time. Though I suspect that conversion is fast. It might add appreciable overhead if you are predicting on very few samples, but the absolute time will be short in those cases (and you could just use the CPU?). We would need to introduce a way to make it easy to convert fitted attributes to Numpy in each estimator to avoid lots of boilerplate. What are other reasons to not do this?

What do others think?

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/112 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
